Add new simulators, search methods, and scoring options

jdramsey opened the pull request:
### Description

This pull request introduces several major updates and features:
- Added new simulators: functional-causal, additive-noise, post-non-linear, and causal-perceptron.
- Implemented Dm/DmFcit latent-discovery search method.
- Introduced new CCA-based scoring options, such as CCA-BIC and BF-LRT.
- Deprecated the MGM algorithm, which has been relocated to the legacy section.

These updates enhance functionality and address contemporary modeling needs.

cg09 commented:
Why send MGM away?

What is CCA?  Are new algorithms documented in the manual?


On Sat, Apr 26, 2025 at 12:33 AM Joseph Ramsey ***@***.***>
wrote:

> Description
>
> This pull request introduces several major updates and features:
>
>    - Added new simulators: functional-causal, additive-noise,
>    post-non-linear, and causal-perceptron.
>    - Implemented Dm/DmFcit latent-discovery search method.
>    - Introduced new CCA-based scoring options, such as CCA-BIC and BF-LRT.
>    - Deprecated the MGM algorithm, which has been relocated to the legacy
>    section.
>
> These updates enhance functionality and address contemporary modeling
> needs.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/cmu-phil/tetrad/pull/1870
> Commit Summary
>
>    - 8433c64
>    <https://github.com/cmu-phil/tetrad/pull/1870/commits/8433c6431e394193eda0940c7de72fe4941f7e0a>
>    Add new simulators, search methods, and scoring options
>
> File Changes
>
> (1 file <https://github.com/cmu-phil/tetrad/pull/1870/files>)
>
>    - *M* tetrad-lib/src/main/resources/docs/manual/index.html
>    <https://github.com/cmu-phil/tetrad/pull/1870/files#diff-5973a6db9aec92b43326f5a8ce9ef2f0790214583f096a24dd3eb9fc02ef6ef1>
>    (114)
>
> Patch Links:
>
>    - https://github.com/cmu-phil/tetrad/pull/1870.patch
>    - https://github.com/cmu-phil/tetrad/pull/1870.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cmu-phil/tetrad/pull/1870>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD4Y3OMDVTK256KWY225Y2D23MEA3AVCNFSM6AAAAAB347D5O2VHI2DSMVQWIX3LMV43ASLTON2WKOZTGAZDCNBXGA4TIMY>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

jdramsey commented:
It became incompatible with the codebase and I couldn't for the life of me
figure out how to update it. The original coders are no longer around.

On Sat, Apr 26, 2025 at 12:47 AM cg09 ***@***.***> wrote:

> *cg09* left a comment (cmu-phil/tetrad#1870)
> <https://github.com/cmu-phil/tetrad/pull/1870#issuecomment-2831844089>
> Why send MGM away? <br> <br> What is CCA? Are new algorithms documented in
> the manual? <br> <br> <br> On Sat, Apr 26, 2025 at 12:33 AM Joseph Ramsey
> ***@***.***&gt; <br> wrote: <br> <br> &gt; Description <br> &gt; <br> &gt;
> This pull request introduces several major updates and features: <br> &gt;
> <br> &gt; - Added new simulators: functional-causal, additive-noise, <br>
> &gt; post-non-linear, and causal-perceptron. <br> &gt; - Implemented
> Dm/DmFcit latent-discovery search method. <br> &gt; - Introduced new
> CCA-based scoring options, such as CCA-BIC and BF-LRT. <br> &gt; -
> Deprecated the MGM algorithm, which has been relocated to the legacy <br>
> &gt; section. <br> &gt; <br> &gt; These updates enhance functionality and
> address contemporary modeling <br> &gt; needs. <br> &gt;
> ------------------------------ <br> &gt; You can view, comment on, or merge
> this pull request online at: <br> &gt; <br> &gt;
> https://github.com/cmu-phil/tetrad/pull/1870 <br> &gt; Commit Summary
> <br> &gt; <br> &gt; - 8433c64 <br> &gt; &lt;
> https://github.com/cmu-phil/tetrad/pull/1870/commits/8433c6431e394193eda0940c7de72fe4941f7e0a&gt;
> <br> &gt; Add new simulators, search methods, and scoring options <br> &gt;
> <br> &gt; File Changes <br> &gt; <br> &gt; (1 file &lt;
> https://github.com/cmu-phil/tetrad/pull/1870/files&gt;) <br> &gt; <br>
> &gt; - *M* tetrad-lib/src/main/resources/docs/manual/index.html <br> &gt;
> &lt;
> https://github.com/cmu-phil/tetrad/pull/1870/files#diff-5973a6db9aec92b43326f5a8ce9ef2f0790214583f096a24dd3eb9fc02ef6ef1&gt;
> <br> &gt; (114) <br> &gt; <br> &gt; Patch Links: <br> &gt; <br> &gt; -
> https://github.com/cmu-phil/tetrad/pull/1870.patch <br> &gt; -
> https://github.com/cmu-phil/tetrad/pull/1870.diff <br> &gt; <br> &gt; —
> <br> &gt; Reply to this email directly, view it on GitHub <br> &gt; &lt;
> https://github.com/cmu-phil/tetrad/pull/1870&gt;, or unsubscribe <br>
> &gt; &lt;
> https://github.com/notifications/unsubscribe-auth/AD4Y3OMDVTK256KWY225Y2D23MEA3AVCNFSM6AAAAAB347D5O2VHI2DSMVQWIX3LMV43ASLTON2WKOZTGAZDCNBXGA4TIMY&gt;
> <br> &gt; . <br> &gt; You are receiving this because you are subscribed to
> this thread.Message <br> &gt; ID: ***@***.***&gt; <br> &gt; <br>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cmu-phil/tetrad/pull/1870#issuecomment-2831844089>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACLFSR35QGMGTQF5L4SUZ6323MFXBAVCNFSM6AAAAAB347D5O2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZRHA2DIMBYHE>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

jdramsey commented:
(I did try. I spent a couple of days on it.)

On Sat, Apr 26, 2025 at 12:52 AM Joseph Ramsey ***@***.***>
wrote:

> It became incompatible with the codebase and I couldn't for the life of me
> figure out how to update it. The original coders are no longer around.
>
> On Sat, Apr 26, 2025 at 12:47 AM cg09 ***@***.***> wrote:
>
>> *cg09* left a comment (cmu-phil/tetrad#1870)
>> <https://github.com/cmu-phil/tetrad/pull/1870#issuecomment-2831844089>
>> Why send MGM away? <br> <br> What is CCA? Are new algorithms documented
>> in the manual? <br> <br> <br> On Sat, Apr 26, 2025 at 12:33 AM Joseph
>> Ramsey ***@***.***&gt; <br> wrote: <br> <br> &gt; Description <br> &gt;
>> <br> &gt; This pull request introduces several major updates and features:
>> <br> &gt; <br> &gt; - Added new simulators: functional-causal,
>> additive-noise, <br> &gt; post-non-linear, and causal-perceptron. <br> &gt;
>> - Implemented Dm/DmFcit latent-discovery search method. <br> &gt; -
>> Introduced new CCA-based scoring options, such as CCA-BIC and BF-LRT. <br>
>> &gt; - Deprecated the MGM algorithm, which has been relocated to the legacy
>> <br> &gt; section. <br> &gt; <br> &gt; These updates enhance functionality
>> and address contemporary modeling <br> &gt; needs. <br> &gt;
>> ------------------------------ <br> &gt; You can view, comment on, or merge
>> this pull request online at: <br> &gt; <br> &gt;
>> https://github.com/cmu-phil/tetrad/pull/1870 <br> &gt; Commit Summary
>> <br> &gt; <br> &gt; - 8433c64 <br> &gt; &lt;
>> https://github.com/cmu-phil/tetrad/pull/1870/commits/8433c6431e394193eda0940c7de72fe4941f7e0a&gt;
>> <br> &gt; Add new simulators, search methods, and scoring options <br> &gt;
>> <br> &gt; File Changes <br> &gt; <br> &gt; (1 file &lt;
>> https://github.com/cmu-phil/tetrad/pull/1870/files&gt;) <br> &gt; <br>
>> &gt; - *M* tetrad-lib/src/main/resources/docs/manual/index.html <br> &gt;
>> &lt;
>> https://github.com/cmu-phil/tetrad/pull/1870/files#diff-5973a6db9aec92b43326f5a8ce9ef2f0790214583f096a24dd3eb9fc02ef6ef1&gt;
>> <br> &gt; (114) <br> &gt; <br> &gt; Patch Links: <br> &gt; <br> &gt; -
>> https://github.com/cmu-phil/tetrad/pull/1870.patch <br> &gt; -
>> https://github.com/cmu-phil/tetrad/pull/1870.diff <br> &gt; <br> &gt; —
>> <br> &gt; Reply to this email directly, view it on GitHub <br> &gt; &lt;
>> https://github.com/cmu-phil/tetrad/pull/1870&gt;, or unsubscribe <br>
>> &gt; &lt;
>> https://github.com/notifications/unsubscribe-auth/AD4Y3OMDVTK256KWY225Y2D23MEA3AVCNFSM6AAAAAB347D5O2VHI2DSMVQWIX3LMV43ASLTON2WKOZTGAZDCNBXGA4TIMY&gt;
>> <br> &gt; . <br> &gt; You are receiving this because you are subscribed to
>> this thread.Message <br> &gt; ID: ***@***.***&gt; <br> &gt; <br>
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/cmu-phil/tetrad/pull/1870#issuecomment-2831844089>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACLFSR35QGMGTQF5L4SUZ6323MFXBAVCNFSM6AAAAAB347D5O2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZRHA2DIMBYHE>
>> .
>> You are receiving this because you modified the open/close state.Message
>> ID: ***@***.***>
>>
>

jdramsey commented:
By the way, it's in a previous version; if anyone wants to use it, they can
download that version and run it. I documented that.

Joe

On Sat, Apr 26, 2025 at 12:52 AM Joseph Ramsey ***@***.***>
wrote:

> (I did try. I spent a couple of days on it.)
>
> On Sat, Apr 26, 2025 at 12:52 AM Joseph Ramsey ***@***.***>
> wrote:
>
>> It became incompatible with the codebase and I couldn't for the life of
>> me figure out how to update it. The original coders are no longer around.
>>
>> On Sat, Apr 26, 2025 at 12:47 AM cg09 ***@***.***> wrote:
>>
>>> *cg09* left a comment (cmu-phil/tetrad#1870)
>>> <https://github.com/cmu-phil/tetrad/pull/1870#issuecomment-2831844089>
>>> Why send MGM away? <br> <br> What is CCA? Are new algorithms documented
>>> in the manual? <br> <br> <br> On Sat, Apr 26, 2025 at 12:33 AM Joseph
>>> Ramsey ***@***.***&gt; <br> wrote: <br> <br> &gt; Description <br> &gt;
>>> <br> &gt; This pull request introduces several major updates and features:
>>> <br> &gt; <br> &gt; - Added new simulators: functional-causal,
>>> additive-noise, <br> &gt; post-non-linear, and causal-perceptron. <br> &gt;
>>> - Implemented Dm/DmFcit latent-discovery search method. <br> &gt; -
>>> Introduced new CCA-based scoring options, such as CCA-BIC and BF-LRT. <br>
>>> &gt; - Deprecated the MGM algorithm, which has been relocated to the legacy
>>> <br> &gt; section. <br> &gt; <br> &gt; These updates enhance functionality
>>> and address contemporary modeling <br> &gt; needs. <br> &gt;
>>> ------------------------------ <br> &gt; You can view, comment on, or merge
>>> this pull request online at: <br> &gt; <br> &gt;
>>> https://github.com/cmu-phil/tetrad/pull/1870 <br> &gt; Commit Summary
>>> <br> &gt; <br> &gt; - 8433c64 <br> &gt; &lt;
>>> https://github.com/cmu-phil/tetrad/pull/1870/commits/8433c6431e394193eda0940c7de72fe4941f7e0a&gt;
>>> <br> &gt; Add new simulators, search methods, and scoring options <br> &gt;
>>> <br> &gt; File Changes <br> &gt; <br> &gt; (1 file &lt;
>>> https://github.com/cmu-phil/tetrad/pull/1870/files&gt;) <br> &gt; <br>
>>> &gt; - *M* tetrad-lib/src/main/resources/docs/manual/index.html <br> &gt;
>>> &lt;
>>> https://github.com/cmu-phil/tetrad/pull/1870/files#diff-5973a6db9aec92b43326f5a8ce9ef2f0790214583f096a24dd3eb9fc02ef6ef1&gt;
>>> <br> &gt; (114) <br> &gt; <br> &gt; Patch Links: <br> &gt; <br> &gt; -
>>> https://github.com/cmu-phil/tetrad/pull/1870.patch <br> &gt; -
>>> https://github.com/cmu-phil/tetrad/pull/1870.diff <br> &gt; <br> &gt; —
>>> <br> &gt; Reply to this email directly, view it on GitHub <br> &gt; &lt;
>>> https://github.com/cmu-phil/tetrad/pull/1870&gt;, or unsubscribe <br>
>>> &gt; &lt;
>>> https://github.com/notifications/unsubscribe-auth/AD4Y3OMDVTK256KWY225Y2D23MEA3AVCNFSM6AAAAAB347D5O2VHI2DSMVQWIX3LMV43ASLTON2WKOZTGAZDCNBXGA4TIMY&gt;
>>> <br> &gt; . <br> &gt; You are receiving this because you are subscribed to
>>> this thread.Message <br> &gt; ID: ***@***.***&gt; <br> &gt; <br>
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/cmu-phil/tetrad/pull/1870#issuecomment-2831844089>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/ACLFSR35QGMGTQF5L4SUZ6323MFXBAVCNFSM6AAAAAB347D5O2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZRHA2DIMBYHE>
>>> .
>>> You are receiving this because you modified the open/close state.Message
>>> ID: ***@***.***>
>>>
>>

cg09 commented:
Never mind the old cold. Input is data and a set of one factor
models. Estimate the values of each latent for each case (Its a standard
problem, I think). Then run a search over the values of the latents.

On Sat, Apr 26, 2025 at 12:53 AM Joseph Ramsey ***@***.***>
wrote:

> *jdramsey* left a comment (cmu-phil/tetrad#1870)
> <https://github.com/cmu-phil/tetrad/pull/1870#issuecomment-2831847209>
> It became incompatible with the codebase and I couldn&#39;t for the life
> of me <br> figure out how to update it. The original coders are no longer
> around. <br> <br> On Sat, Apr 26, 2025 at 12:47 AM cg09 ***@***.***&gt;
> wrote: <br> <br> &gt; *cg09* left a comment (cmu-phil/tetrad#1870) <br>
> &gt; &lt;
> https://github.com/cmu-phil/tetrad/pull/1870#issuecomment-2831844089&gt;
> <br> &gt; Why send MGM away? &lt;br&gt; &lt;br&gt; What is CCA? Are new
> algorithms documented in <br> &gt; the manual? &lt;br&gt; &lt;br&gt;
> &lt;br&gt; On Sat, Apr 26, 2025 at 12:33 AM Joseph Ramsey <br> &gt;
> ***@***.***&amp;gt; &lt;br&gt; wrote: &lt;br&gt; &lt;br&gt; &amp;gt;
> Description &lt;br&gt; &amp;gt; &lt;br&gt; &amp;gt; <br> &gt; This pull
> request introduces several major updates and features: &lt;br&gt; &amp;gt;
> <br> &gt; &lt;br&gt; &amp;gt; - Added new simulators: functional-causal,
> additive-noise, &lt;br&gt; <br> &gt; &amp;gt; post-non-linear, and
> causal-perceptron. &lt;br&gt; &amp;gt; - Implemented <br> &gt; Dm/DmFcit
> latent-discovery search method. &lt;br&gt; &amp;gt; - Introduced new <br>
> &gt; CCA-based scoring options, such as CCA-BIC and BF-LRT. &lt;br&gt;
> &amp;gt; - <br> &gt; Deprecated the MGM algorithm, which has been relocated
> to the legacy &lt;br&gt; <br> &gt; &amp;gt; section. &lt;br&gt; &amp;gt;
> &lt;br&gt; &amp;gt; These updates enhance functionality and <br> &gt;
> address contemporary modeling &lt;br&gt; &amp;gt; needs. &lt;br&gt;
> &amp;gt; <br> &gt; ------------------------------ &lt;br&gt; &amp;gt; You
> can view, comment on, or merge <br> &gt; this pull request online at:
> &lt;br&gt; &amp;gt; &lt;br&gt; &amp;gt; <br> &gt;
> https://github.com/cmu-phil/tetrad/pull/1870 &lt;br&gt; &amp;gt; Commit
> Summary <br> &gt; &lt;br&gt; &amp;gt; &lt;br&gt; &amp;gt; - 8433c64
> &lt;br&gt; &amp;gt; &amp;lt; <br> &gt;
> https://github.com/cmu-phil/tetrad/pull/1870/commits/8433c6431e394193eda0940c7de72fe4941f7e0a&amp;gt;
> <br> &gt; &lt;br&gt; &amp;gt; Add new simulators, search methods, and
> scoring options &lt;br&gt; &amp;gt; <br> &gt; &lt;br&gt; &amp;gt; File
> Changes &lt;br&gt; &amp;gt; &lt;br&gt; &amp;gt; (1 file &amp;lt; <br> &gt;
> https://github.com/cmu-phil/tetrad/pull/1870/files&amp;gt;) &lt;br&gt;
> &amp;gt; &lt;br&gt; <br> &gt; &amp;gt; - *M*
> tetrad-lib/src/main/resources/docs/manual/index.html &lt;br&gt; &amp;gt;
> <br> &gt; &amp;lt; <br> &gt;
> https://github.com/cmu-phil/tetrad/pull/1870/files#diff-5973a6db9aec92b43326f5a8ce9ef2f0790214583f096a24dd3eb9fc02ef6ef1&amp;gt;
> <br> &gt; &lt;br&gt; &amp;gt; (114) &lt;br&gt; &amp;gt; &lt;br&gt; &amp;gt;
> Patch Links: &lt;br&gt; &amp;gt; &lt;br&gt; &amp;gt; - <br> &gt;
> https://github.com/cmu-phil/tetrad/pull/1870.patch &lt;br&gt; &amp;gt; -
> <br> &gt; https://github.com/cmu-phil/tetrad/pull/1870.diff &lt;br&gt;
> &amp;gt; &lt;br&gt; &amp;gt; — <br> &gt; &lt;br&gt; &amp;gt; Reply to this
> email directly, view it on GitHub &lt;br&gt; &amp;gt; &amp;lt; <br> &gt;
> https://github.com/cmu-phil/tetrad/pull/1870&amp;gt;, or unsubscribe
> &lt;br&gt; <br> &gt; &amp;gt; &amp;lt; <br> &gt;
> https://github.com/notifications/unsubscribe-auth/AD4Y3OMDVTK256KWY225Y2D23MEA3AVCNFSM6AAAAAB347D5O2VHI2DSMVQWIX3LMV43ASLTON2WKOZTGAZDCNBXGA4TIMY&amp;gt;
> <br> &gt; &lt;br&gt; &amp;gt; . &lt;br&gt; &amp;gt; You are receiving this
> because you are subscribed to <br> &gt; this thread.Message &lt;br&gt;
> &amp;gt; ID: ***@***.***&amp;gt; &lt;br&gt; &amp;gt; &lt;br&gt; <br> &gt;
> <br> &gt; — <br> &gt; Reply to this email directly, view it on GitHub <br>
> &gt; &lt;
> https://github.com/cmu-phil/tetrad/pull/1870#issuecomment-2831844089&gt;,
> <br> &gt; or unsubscribe <br> &gt; &lt;
> https://github.com/notifications/unsubscribe-auth/ACLFSR35QGMGTQF5L4SUZ6323MFXBAVCNFSM6AAAAAB347D5O2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZRHA2DIMBYHE&gt;
> <br> &gt; . <br> &gt; You are receiving this because you modified the
> open/close state.Message <br> &gt; ID: ***@***.***&gt; <br> &gt; <br>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cmu-phil/tetrad/pull/1870#issuecomment-2831847209>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD4Y3OOTZPNQKSF34YZA6A323MGLTAVCNFSM6AAAAAB347D5O2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZRHA2DOMRQHE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

jdramsey commented:
It's actually not a standard problem. The place where I got stuck was where a complicated Matlab function was clumsily translated into Java. That's what I couldn't figure out how update; the matrix algebra there was fragile.

jdramsey commented:
Yeah that's what the issue was. We updated out matrix library because the old one had fallen out of repair and was not longer supported. The Matlab translation of that search could not be translated into the new matrix algebra class. Everythign else in Tetrad could be. In any case, if someone wants to use it they can, just not in the most recent version.

cg09 commented:
Here is Chat gpt code for the problem using principal components:
import org.apache.commons.math3.linear.*;
import java.util.Arrays;

public class PCAFactorApproximation {
    public static void main(String[] args) {
        // Example data: rows = cases, columns = observed variables
        double[][] data = {
            {2.5, 2.4},
            {0.5, 0.7},
            {2.2, 2.9},
            {1.9, 2.2},
            {3.1, 3.0},
            {2.3, 2.7},
            {2, 1.6},
            {1, 1.1},
            {1.5, 1.6},
            {1.1, 0.9}
        };

        RealMatrix X = MatrixUtils.createRealMatrix(data);
        RealMatrix XCentered = centerColumns(X);

        // Compute covariance matrix
        RealMatrix covariance = new
Covariance(XCentered).getCovarianceMatrix();

        // Eigen decomposition
        EigenDecomposition eig = new EigenDecomposition(covariance);
        RealMatrix eigenvectors = eig.getV();
        RealVector firstPC = eigenvectors.getColumnVector(0);

        // Project data onto first principal component
        double[] latentScores = XCentered.operate(firstPC).toArray();

        // Output estimated latent variable (first principal component)
        System.out.println("Estimated latent variable scores:");
        Arrays.stream(latentScores).forEach(score ->
System.out.printf("%.4f%n", score));
    }

    // Helper: center the columns (zero mean)
    private static RealMatrix centerColumns(RealMatrix matrix) {
        int rows = matrix.getRowDimension();
        int cols = matrix.getColumnDimension();
        double[][] centered = new double[rows][cols];

        for (int j = 0; j < cols; j++) {
            double mean =
Arrays.stream(matrix.getColumn(j)).average().orElse(0);
            for (int i = 0; i < rows; i++) {
                centered[i][j] = matrix.getEntry(i, j) - mean;
            }
        }
        return MatrixUtils.createRealMatrix(centered);
    }
}

On Sat, Apr 26, 2025 at 1:13 AM Joseph Ramsey ***@***.***>
wrote:

> *jdramsey* left a comment (cmu-phil/tetrad#1870)
> <https://github.com/cmu-phil/tetrad/pull/1870#issuecomment-2831860701>
>
> It's actually not a standard problem. The place where I got stuck was
> where a complicated Matlab function was clumsily translated into Java.
> That's what I couldn't figure out how update; the matrix algebra there was
> fragile.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cmu-phil/tetrad/pull/1870#issuecomment-2831860701>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD4Y3OJSTFWBWFAL2Q22B5323MIX3AVCNFSM6AAAAAB347D5O2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZRHA3DANZQGE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

jdramsey commented:
Principle components was the issue. There was an optimization of some sort. I couldn't follow the code. I can dig up the code and send it to you if you want to take a shot at it.
